examples/hamiltonian_representation.py

```diff
@@ -0,0 +1,135 @@
+    def __mul__(self, other):
+        hamiltonians = {}
+        for h1, w1 in self.hamiltonians.items():
+            for h2, w2 in other.hamiltonians.items():
+                h = tuple(sorted(set(h1).symmetric_difference(h2)))
+                w = w1 * w2
+                if h not in hamiltonians:
+                    hamiltonians[h] = 0
+                hamiltonians[h] += w
+        return HamiltonianList(hamiltonians)
+
+    @staticmethod
+    def O():
+        return HamiltonianList({})
+
+    @staticmethod
+    def I():
+        return HamiltonianList({(): 1.0})
+
+    @staticmethod
+    def Z(i: int):
+        return HamiltonianList({(i,): 1.0})
+
+
+def build_hamiltonian_from_boolean(boolean_expr, name_to_id) -> HamiltonianList:
+    """Builds the Hamiltonian representation of Boolean expression as per [1]:
+
+    Args:
+        boolean_expr: A Sympy expression containing symbols and Boolean operations
+        name_to_id: A dictionary from symbol name to an integer, typically built by calling
+            get_name_to_id().
+
+    Return:
+        The HamiltonianList that represents the Boolean expression.
+    """
+
+    if isinstance(boolean_expr, (And, Not, Or, Xor)):
+        sub_hamiltonians = [
+            build_hamiltonian_from_boolean(sub_boolean_expr, name_to_id)
+            for sub_boolean_expr in boolean_expr.args
+        ]
+        # We apply the equalities of theorem 1 of [1].
+        if isinstance(boolean_expr, And):
+            hamiltonian = HamiltonianList.I()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Not):
+            assert len(sub_hamiltonians) == 1
+            hamiltonian = HamiltonianList.I() - sub_hamiltonians[0]
+        elif isinstance(boolean_expr, Or):
+            hamiltonian = HamiltonianList.O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - hamiltonian * sub_hamiltonian
+        elif isinstance(boolean_expr, Xor):
+            hamiltonian = HamiltonianList.O()
+            for sub_hamiltonian in sub_hamiltonians:
+                hamiltonian = hamiltonian + sub_hamiltonian - 2.0 * hamiltonian * sub_hamiltonian
+        return hamiltonian
+    elif isinstance(boolean_expr, Symbol):
+        # Table 1 of [1], entry for 'x' is '1/2.I - 1/2.Z'
+        i = name_to_id[boolean_expr.name]
+        return 0.5 * HamiltonianList.I() - 0.5 * HamiltonianList.Z(i)
+    else:
+        raise ValueError(f'Unsupported type: {type(boolean_expr)}')
+
+
+def get_name_to_id(boolean_exprs):
+    # For run-to-run identicalness, we sort the symbol name lexicographically.
+    symbol_names = sorted(
+        {symbol.name for boolean_expr in boolean_exprs for symbol in boolean_expr.free_symbols}
+    )
+    return {symbol_name: i for i, symbol_name in enumerate(symbol_names)}
+
+
+def build_circuit_from_hamiltonians(hamiltonians, qubits, theta):
```

Comment:
Do you think it is good to have another function `build_circuit_from_boolean_expressions`, wrapping `parse_expr`, `build_hamiltonian_from_boolean` and `get_name_to_id`?  I know it is NOT useful in QAOA since the same `hamiltonians` with different theta will be used multiple times. But what about potential cases?

Reply:
Done.